src/or/control.c

```diff
@@ -5794,88 +5794,459 @@ control_event_bandwidth_used(uint32_t n_read, uint32_t n_written)
+static int
+privcount_is_exit(const edge_connection_t* exitconn,
+                  const or_circuit_t *orcirc)
+{
+  if (!exitconn) {
+    /* Assume missing connections are *not* Exit connections */
+    return 0;
+  }
+
+  const or_circuit_t* oc = privcount_get_const_or_circuit(exitconn, orcirc);
+
+  if (!oc) {
+    /* Assume missing circuits are *not* OR circuits */
+    return 0;
+  }
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (CIRCUIT_IS_ORIGIN(TO_CIRCUIT(oc))) {
+    return 0;
+  }
+
+  if (!oc->p_chan) {
+    return 0;
+  }
+
+  return TO_CONN(exitconn)->type == CONN_TYPE_EXIT;
+}
+
+/* Should PrivCount ignore events from this connection or circuit?
+ * Directory connections and DNS resolves are considered overhead, so this
+ * function is not suitabe for filtering EVENT_PRIVCOUNT_DNS_RESOLVED events.
+ */
+static int
+privcount_is_overhead(const connection_t* conn, const circuit_t *circ)
+{
+#if PRIVCOUNT_DNS_RESOLVE_IS_OVERHEAD
+  if (privcount_conn_is_dns_resolve(conn)) {
+    return 1;
+  }
+#endif
+
+  /* if the circuit started here, this is our own stream and we can ignore it
+   */
+  if (circ && CIRCUIT_IS_ORIGIN(circ)) {
+    return 1;
+  }
+
+  return privcount_is_dir(conn, circ);
+}
+
+/* Should PrivCount count streams and bytes from this connection or circuit?
+ */
+int
+privcount_is_counted_for_bytes(const edge_connection_t* exitconn,
+                               const or_circuit_t *orcirc)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_BYTES_TRANSFERRED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_STREAM_ENDED)) {
+    return 0;
+  }
+
+  /* Ignore events with missing connections */
+  if (!exitconn) {
+    return 0;
+  }
+
+  if (privcount_is_overhead(TO_CONN(exitconn), TO_CIRCUIT(orcirc))) {
+    return 0;
+  }
+
+  /* Bytes are only counted at exit connections */
+  return privcount_is_exit(exitconn, orcirc);
+}
+
+/* Should PrivCount ignore cells, circuits, and connections from this circuit?
+ */
+int
+privcount_is_counted_for_cells(const connection_t *conn,
+                               const circuit_t *circ)
+{
+  /* Ignore events that are disabled */
+  if (!get_options()->EnablePrivCount) {
+    return 0;
+  }
+
+  if (!EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CIRCUIT_ENDED) &&
+      !EVENT_IS_INTERESTING(EVENT_PRIVCOUNT_CONNECTION_ENDED)) {
+    return 0;
+  }
+
+  /* Ignore events which are missing both connection and circuit */
+  if (!conn && !circ) {
+    return 0;
+  }
+
+  if (privcount_is_overhead(conn, circ)) {
+    return 0;
+  }
+
+  /* Cells get counted at non-exit connections as well as exit connections */
+  return 1;
+}
+
+/* Should PrivCount ignore DNS requests from this connection or circuit?
+ */
+int
+privcount_is_counted_for_dns(const edge_connection_t* exitconn,
```

Comment:
I fixed this, see my comment above in https://github.com/privcount/tor/pull/4#discussion_r113629202